test/parallel/test-http2-client-window-size.js

```diff
@@ -0,0 +1,104 @@
+const common = require('../common');
+const h2 = require('http2');
+
+// Helper method which takes an array of buffers and returns a buffer containing
+// all the bytes in each buffer in order
+function concatBuffers(buffers) {
+  const resultLength =
+    buffers.reduce((prev, buf) => prev + buf.length, 0);
+  const result = Buffer.alloc(resultLength);
+  let offset = 0;
+  for (let i = 0; i < buffers.length; i++) {
+    buffers[i].copy(result, offset);
+    offset += buffers[i].length;
+  }
+  return result;
+}
+
+// Given a list of buffers and an initial window size, have a server write
+// each buffer to the HTTP2 Writable stream, and let the client verify that
+// all of the bytes were sent correctly
+function run(buffers, initialWindowSize) {
+  const expectedBuffer = concatBuffers(buffers);
+
+  const server = h2.createServer();
+  const a = [];
+  server.on('stream', (stream, headers, flags) => {
+    let p = Promise.resolve();
+    for (let i = 0; i < buffers.length; i++) {
+      p = p.then(() => new Promise(resolve => {
+        const drained = stream.write(buffers[i], 'buffer', resolve);
+        if (!drained) {
+          stream.once('drain', resolve);
+        }
+      }));
+    }
+    p = p.then(() => {
+      stream.end();
+    });
+  });
+  server.listen(0);
+
+  server.on('listening', common.mustCall(function() {
+    const port = this.address().port;
+
+    const client =
+      h2.connect({
+        authority: 'localhost',
+        protocol: 'http:',
+        port
+      }, {
+        settings: {
+          initialWindowSize
+        }
+      }).on('connect', common.mustCall(() => {
+        const req = client.request({
+          ':method': 'GET',
+          ':path': '/'
+        });
+        let actualBuffer = Buffer.alloc(expectedBuffer.length);
+        let offset = 0;
+        req.on('data', data => {
+          data.copy(actualBuffer, offset);
+          offset += data.length;
+        });
+        req.on('end', common.mustCall(() => {
+          assert.strictEqual(offset, actualBuffer.length);
+          assert.strictEqual(Buffer.compare(actualBuffer, expectedBuffer), 0);
+          // shut down
+          client.socket.destroy();
+          server.close();
+        }));
+        req.end();
+      }));
+  }));
+}
+
+const bufferValueRange = [0, 1, 2, 3];
+const buffersList = [
+  bufferValueRange.map(a => Buffer.alloc(1 << 4).fill(a)),
+  bufferValueRange.map(a => Buffer.alloc((1 << 8) - 1).fill(a)),
+  bufferValueRange.map(a => Buffer.alloc(1 << 8).fill(a)),
+  bufferValueRange.map(a => Buffer.alloc(1 << 17).fill(a))
+];
+const initialWindowSizeList = [
+  1 << 4,
+  (1 << 8) - 1,
+  1 << 8,
+  1 << 17,
+  undefined, // use default window size which is (1 << 16) - 1
+];
+
+// Call `run` on each element in the cartesian product of buffersList and
+// initialWindowSizeList.
+for (const buffers of buffersList) {
+  for (const initialWindowSize of initialWindowSizeList) {
+    run(buffers, initialWindowSize);
+  }
+}
```

Comment:
Sure thing -- though, on that note, this test takes a few seconds to run, and running these sequentially would probably take even longer. Do you have any suggestions on how this could be cut down? And whether it's necessary to do so?